### ISSUE TITLE
why does nniFromAny not recognize an integer argument for a keyed dimension?

### ISSUE THREAD

**Issue by lisp** [gjmcn/data-cube#5]
if a key map is defined, an attempt it address the respective dimension with an integer index yields an error.
is there some reason for this constraint?

**Comment by gjmcn** [comment]
A dimension can have indices or keys, but not both &mdash; keys can be integers, so it would not be clear in e.g. `x.row(2)` if `2` was an index or a key.

Note that `x.row(2)` will work when there are row keys, if there is a row with key `2` (which need not be the third row).

**Comment by lisp** [comment]
> A dimension can have indices or keys

why? i understand your example, but that is not a reason the prevent a use case where the namespaces are distinct.

**Comment by gjmcn** [comment]
Not sure I follow. Do you mean that if `x` has row keys then `x.row(2)` should look for a key `2`, but if there is no such key, should use `2` as an index (i.e. get the third row)?

**Comment by lisp** [comment]
in any case, something less counter-intuitive than examining an integer index, determining that it is not a valid key and throwing an error to the effect that an integer was expected, because the map returned none.
more transparent would be to be able to declare the behaviour, perhaps by declaring the map's domain.

**Comment by computermacgyver** [comment]
There are some really strange behaviours that occur in Jupyter notebooks when integers are used as labels. Jupyter solves this by having separate methods [`iloc`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.iloc.html) to look up by integer position and [`loc`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.loc.html#pandas.DataFrame.loc) to look up by labels. It still does risk some nonintuitive behaviour when a method that accepts either a label or integer position, but it is an option that could be considered here.

In short, I think there are good reasons why the current design of data-cube doesn't allow labels and integer-positions to be mixed, but longer term adding additional methods to deal with such a situation could be nice.

**Comment by lisp** [comment]
the use case involves creating a cube from data which is readily associated with numeric locations, but using it with symbolic locations. yes, there are appropriate approaches:
- map the locations to keys to set the entries
- transform the initial data into a vector
- add the keys after filling the cube

but none is a clear as copying by location.

**Comment by gjmcn** [comment]
In general, I agree with @computermacgyver that additional methods that use indices regardless of whether a dimension has keys or not would be the way to go.

For the current behavior:

> in any case, something less counter-intuitive than examining an integer index, determining that it is not a valid key and throwing an error to the effect that an integer was expected, because the map returned none.

The "number expected" error message is very confusing and should be changed.
